### ISSUE TITLE
PPLNS shares on dashboard stop updating when there are no active miners

### ISSUE THREAD

**Issue by Venthos** [Venthos/nodejs-pool#1]
Since nodejs-pool only updates miner statistics when there is an active worker for that address, the PPLNS Shares reported to the UI "freeze" at their last known value when there are no active workers.  This causes the UI to perpetually show incorrect PPLNS related information until at least one worker becomes active again.

While I could simply "zero" out PPLNS Shares once there are no more active workers, that is not how PPLNS operates.  An ideal situation is to find some way to sanely advance the statistics of recently inactive miners (less than 24 hours since last hash?) as if they were active.  This would accurately portray the "slow roll off" of PPLNS shares as they start advancing past the PPLNS depth.

**Comment by bobbieltd** [comment]
So your eatimation will be incorrect when the miners stop mining ? Hix

**Comment by bobbieltd** [comment]
What if he stops mining and block found, he still got the same est. rewards without any new shares ?

**Comment by bobbieltd** [comment]
My suggestion is that we won’t calculate any PPLNS shares (because of high network diffuculty deviation). Worker will calculate directly est. rewards (reward based on shares and difficulty at that precise momeny).

**Comment by Venthos** [comment]
This is only a visual bug on the dashboard.  It does not affect payouts in anyway.  Payouts are handled by blockManager separately with its own logic.

The reason this happens is that the entire logic loop that updates hashrates, pplnsShares, etc. for miner stats only runs if there is an active worker.  It gathers a list of active workers first, then processes all the updates for them.

It does this because it would be a waste of time/resources to "update" stats for every worker that ever mined on the pool since the beginning of time, even if they haven't hashed at the pool for a month.

The fix here is just to break out another logical loop that iterates through workers that are not active, but have a positive pplnsShare value, and update that value.  Or, if there's a sane spot to simply "add on" those workers to the list that gets processed by the main loop.  I just need to set aside time to implementing the fix.

**Comment by Venthos** [comment]
I think I have it fixed now.  It ended up being a super simple fix after I took the time to step through the existing logic.  However, it will take a bit for me to test it since I need to mine for a period of time, stop mining, and observe the slow roll down in stats.  After confirming it works, I'll commit the change.

This particular feature just exposes PPLNS Shares to the API, so you definitely could use this in some other way than calculating estimated rewards.  I just choose to do it that way.

You are correct though that the way that PPLNS works is that if I stop mining, and a block is found after I stopped entirely, I still may get some rewards if I still have shares within the PPLNS depth.  That is intentionally how that payout method works.  It pays you based on how many shares are yours within the past (N * currentDifficulty) worth of shares for the pool.
